src/index.cgi.go

```diff
@@ -0,0 +1,175 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"net/http"
+	"net/http/cgi"
+	"encoding/json"
+	"bytes"
+	"log"
+	"strings"
+)
+
+func main() {
+	if err := cgi.Serve(http.HandlerFunc(cgiHandler)); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func cgiHandler(w http.ResponseWriter, r *http.Request) {
+	if synoTokenRedirect(w, r) {
+		return
+	}
+
+	user, err := auth()
+	if err != nil {
+		http.Error(w, err.Error(), 403)
+		log.Fatal("tag 1: %v", err.Error())
+		return
+	}
+
+	s, err := tailscaleStatus()
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		log.Fatal("tag 2: %v", err.Error())
+		return
+	}
+	if uid := os.Getuid() ; uid != 0 {
+		http.Error(w, fmt.Sprintf("cgi running as %d, not root", uid), 500)
+		return
+	}
+
+	buf := new(bytes.Buffer)
+	wr := func(format string, args ...interface{}) {
+		for i, arg := range args {
+			v := fmt.Sprintf("%v", arg)
+			v = strings.Replace(v, "&", "&amp;", -1)
+			args[i] = strings.Replace(v, "<", "&lt;", -1)
+		}
+		fmt.Fprintf(buf, format+"\n", args...)
+	}
+	wr("<!doctype html>")
+	wr("<html><title>Tailscale Synology App</title><body>")
+	wr("<h1>Tailscale</h1>")
+	wr(`<div style="float:right;">%s</div>`, user)
+	wr("<table>")
+	wr("<tr><th>Status:</th><td>%s</td></tr>", s.BackendState)
+	wr("<tr><th>Device Name:</th><td>%s</td></tr>", s.Self.DNSName)
+	wr("<tr><th>Tailscale IPs:</th><td>")
+	for i, ip := range s.TailscaleIPs {
+		if i > 0 {
+			wr(", ")
+		}
+		wr("%s", ip)
+	}
+	wr("</td></tr>")
+	wr("</table>")
+
+	wr(`<form method="POST">`)
+	wr(`<p><input type="submit" value="Log in…"></p>`)
+	wr(`</form>`)
+	url := s.AuthURL
+	if r.FormValue("action") == "login" {
+		_, url, err = tailscaleUp()
+		if err != nil {
+			wr(`<p style="color: red; font-weight: bold;">%v</p>`, err)
+		}
+		if url != "" {
+			s.AuthURL = url
+		}
+	}
+	if s.BackendState != "Running" && url != "" {
+		wr(`<p><b>To authorize this device visit: <a href=%q>%s</a></b></p>`, url, url)
+	}
+
+	wr("</body></html>")
+
+	w.Write(buf.Bytes())
+}
+
+type status struct {
```

Comment:
if we moved this CGI into the main tree, we could just use the right type easily and they wouldn't get out of sync.

Reply:
what about making this tool be part of `cmd/tailscale` when `+build linux`, as an undocumented `tailscale syno-cgi` subcommand. Then we could hard link it into two places as `foo.cgi` and make cmd/tailscale go into CGI mode if its os.Argv[0] is `foo.cgi`.